File: C++/src/benchmarks/source/DataSerialization.cpp
//
// Created by saeed on 11/4/19.
//


#include <flatbuffers/TweetStatusFlatBuffers.h>
#include "DataSerialization.h"

using namespace std;

int main(int argc, char *argv[]) {

    if (argc < 3) {
        cout << "Requires Tweeter JSON file path!!" << endl;
        cout << "Requires Data Serialization Type:" << endl;
        cout << " 1: HandCoded Data serialization" << endl;
        cout << " 2: InPlace Data Serialization" << endl;
        cout << " 3: Boost Data Serialization" << endl;
        cout << " 4: ProtoData Data Serialization" << endl;
        cout << "Run as: ./main <path/to/tweet.json> <serializationtype> <path/to/output/file>" << endl;
        return 0;
    }

    // Set input file path:
    string inputFile = argv[1];

    int serializationType = atoi(argv[2]);

    //Set output file path
    string outFile = argv[3];

    //set number Of Tweets
    long numberOfTweets=atoi(argv[4]);
    int round=atoi(argv[5]);

    bool taskset;
    int ts=atoi(argv[6]);
    if (ts==1)
        taskset= true;
    else
        taskset= false;

    switch (serializationType) {
        case 1:
        case 3:
        case 5:{
            DataSerialization<TweetStatus> *dataSerialization=new DataSerialization<TweetStatus>(inputFile,outFile,serializationType,numberOfTweets,round,taskset);
            dataSerialization->runTheDataSerialization();

            delete dataSerialization;
            break;
        }

        case 2: {
            DataSerialization<TweetStatusIP> *dataSerialization=new DataSerialization<TweetStatusIP>(inputFile,outFile,serializationType,numberOfTweets,round,taskset);
            dataSerialization->runTheDataSerialization();

            delete dataSerialization;
            break;
        }

        case 4: {
            DataSerialization<TweetStatusProto> *dataSerialization=new DataSerialization<TweetStatusProto>(inputFile,outFile,serializationType,numberOfTweets,round,taskset);
            dataSerialization->runTheDataSerialization();

            delete dataSerialization;
            break;
        }
        case 6: {
            DataSerialization<TweetStatusFlatBuffers> *dataSerialization=new DataSerialization<TweetStatusFlatBuffers>(inputFile,outFile,serializationType,numberOfTweets,round,taskset);
            dataSerialization->runTheDataSerialization();

            delete dataSerialization;
            break;
        }
    }
    return 0;

}

File: Java/src/main/java/at/tugraz/benchmarks/DataSerialization.java
package at.tugraz.benchmarks;

import com.google.gson.Gson;
import at.tugraz.filehandler.FileHandler;
import at.tugraz.filehandler.LogFileHandler;
import at.tugraz.filehandler.ReadTweet;
import at.tugraz.tweet.TweetStatus;
import org.apache.log4j.PropertyConfigurator;
import java.io.IOException;
import java.util.Properties;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public class DataSerialization {

	public static void main(String[] args) throws IOException {

		Properties prop = new Properties();

		PropertyConfigurator.configure(prop);
		prop.setProperty("log4j.appender.WORKLOG.File", "log4j.properties");

		PropertyConfigurator.configure("log4j.properties");

		// Set input file path:
		String inputFile = args[0];

		int serializationType = Integer.parseInt(args[1]);

		//Set output file path
		String outFile = args[2];

		// set number of tweets:
		long numberOfTweets = Long.parseLong(args[3]);

		int round = Integer.parseInt(args[4]);

		boolean taskset = Boolean.parseBoolean(args[5]);

		String logFileName = "bin/benchmark/writeobjects/result_java_writeobjects_" + numberOfTweets + "_" + round + ".txt";

		FileHandler fileHandler = new FileHandler(outFile, serializationType);
		fileHandler.prepareToWrite();

		BlockingQueue<String> queue = new ArrayBlockingQueue<>(1000);
		ReadTweet readTweet = new ReadTweet(queue, inputFile);
		Thread rt = new Thread(readTweet);
		rt.start();
		//starting consumer to consume raw tweet from queue

		double elapsedSeconds = 0;

		try {
			String line;
			readTweet.setReadStatus(true);
			for(long i = 0; i < numberOfTweets; i++) {
				line = queue.take();
				Gson gson = new Gson();
				TweetStatus tweetStatus = gson.fromJson(line, TweetStatus.class);
				long tmpTime = System.nanoTime();
				fileHandler.appendObjectToFile(tweetStatus);
				elapsedSeconds += (System.nanoTime() - tmpTime);
			}
			readTweet.setReadStatus(false);
			queue.take();
		}
		catch(Exception e) {
			e.printStackTrace();
		}
		fileHandler.appendObjectToFileFlush();

		// Time Calculation
		elapsedSeconds /= 1000000000.0;

		LogFileHandler logFileHandler = new LogFileHandler(logFileName);
		logFileHandler
			.addLog(false, serializationType, true, "TweetStatus", fileHandler.getIoTime(), elapsedSeconds, taskset);

	}
}